sycl/doc/extensions/proposed/sycl_ext_oneapi_barrier.asciidoc

```diff
@@ -0,0 +1,251 @@
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+This extension defines a `sycl::ext::oneapi::experimental::barrier` class
+aligned with the `std::barrier` class defined in {cpp}20. This extension may
+also be considered a generalization of the existing
+link:../experimental/sycl_ext_oneapi_cuda_async_barrier.asciidoc[sycl_ext_oneapi_cuda_async_barrier]
+extension for non-CUDA backends.
+
+`sycl::ext::oneapi::experimental::barrier` is _not_ a general replacement for
+`sycl::group_barrier`, and is intended for usage by experts seeking to
+synchronize subsets of threads of execution that do not necessarily correspond
+to groups of work-items.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_BARRIER` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== `barrier` class
+
+The `sycl::ext::oneapi::experimental::barrier` class has the same interface as
+the `std::barrier` class from {cpp}20, with an additional `Scope` template
+parameter denoting a memory scope broad enough to contain all threads of
+execution that are able to use the barrier.
+
+[NOTE]
+====
+The threads of execution that are synchronized by this barrier may be a
+combination of host threads and device threads (i.e. work-items), if the
+barrier is visible to all threads of execution and has a sufficiently broad
+scope.
+====
+
+This extension specification provides a brief overview of the functionality of
+the `barrier` class. If there are any differences in the definitions of this
+functionality, the wording in the {cpp}20 specification takes precedence.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  template <sycl::memory_scope Scope, class CompletionFunction = __unspecified__>
+  class barrier {
+  public:
+    using arrival_token = __unspecified__;
+
+    static constexpr ptrdiff_t max() noexcept;
+
+    constexpr explicit barrier(ptrdiff_t expected,
```

Comment:
I don't think so.  The way I have imagined this being used, the barrier itself is allocated in memory somewhere.  I think you're envisioning a usage model like:

```c++
q.parallel_for(..., [=](sycl::nd_item it) {
  syclex::barrier<sycl::memory_scope::device> mybar(nthreads);
});
```

...but I don't think this is aligned with `std::barrier`.  The above would create a separate `syclex::barrier` for every work-item, and store those barriers in private memory.  To be consistent with the usage model of `std::barrier`, I think we would want people to write something similar to the below:

```c++

syclex::barrier<sycl::memory_scope::device>* mybar;
q.parallel_for(..., [=](sycl::nd_item it) {

  if (/* This work-item was elected by the user to initialize the barrier*/) {
    *mbar = syclex::barrier<sycl::memory_scope::device>(nthreads);
  }
  
  // Synchronize the work-items via some existing mechanism, to ensure that mbar has been initialized
  sycl::group_barrier(it.get_root_group());

  // Use the barrier
  mbar->arrive_and_wait();
}
```

I say "similar" because I think we need some implementation experience here to help us figure out what these patterns look like, and what the common cases are.  I'm not sure whether the above is safe or if we need to use placement new.  I'm also not sure if it'll be more common for kernels to allocate global barriers inside of the kernels that use them, or if they're more likely to use `device_global` or perform the initialization in a `single_task`.

Reply:
Having looked at this again this morning, I've convinced myself that it should indeed be using placement new.

Reply:
Could you construct the `barrier` object on the host in USM shared memory, and then use the `barrier` in a kernel to synchronize work-items?  That seems like it would be convenient, and I don't see anything in this spec that says this is illegal.

As you note, it would also be possible to construct the `barrier` in USM memory (any type) from a `single_task` kernel.  Then you could use the `barrier` in other kernels to synchronize work-items.

Both of these options would require placement-new to construct the `barrier`.

Hmmm .... do we need to say that `barrier` objects can only be constructed in memory that supports atomic operations?  For example, is it legal to construct a `barrier` in USM host memory if `usm_atomic_host_allocations` is false?

Reply:
> Could you construct the barrier object on the host in USM shared memory, and then use the barrier in a kernel to synchronize work-items? That seems like it would be convenient, and I don't see anything in this spec that says this is illegal.

Yes, I think this should be legal.

> Hmmm .... do we need to say that barrier objects can only be constructed in memory that supports atomic operations? For example, is it legal to construct a barrier in USM host memory if usm_atomic_host_allocations is false?

I think this is tied to the question in the other comment regarding the atomic queries.  I believe the following statements are true:
- C++20 barriers provide a _strongly happens before_ guarantee, so have similar synchronization behavior to atomics/fences using `seq_cst`.
- All threads of execution have to be able to access the barrier object concurrently (regardless of whether it builds on atomic operations).
- It should always be possible to implement barrier using atomics and fences (if those atomics and fences support `seq_cst` at the required scope).

Given that, perhaps we need to say that:
- Barriers can only be constructed in memory that supports atomic operations.
- Barriers can only be used on devices that support `seq_cst` atomics and fences.
- Barriers at system scope can only be constructed in memory that supports concurrent atomic operations from all threads of execution using the barrier (which may include threads on other devices and/or the host).

Although I agree with your comment that some devices could implement support for barriers natively, given their strong ordering requirements it seems unlikely that a device would have hardware capable of satisfying sequential consistency for `std::barrier`, but would still be unable to support sequentially consistent atomics and fences.

Reply:
> I don't think so. The way I have imagined this being used, the barrier itself is allocated in memory somewhere. I think you're envisioning a usage model like:
> 
> ```c++
> q.parallel_for(..., [=](sycl::nd_item it) {
>   syclex::barrier<sycl::memory_scope::device> mybar(nthreads);
> });
> ```
> 
> ...but I don't think this is aligned with `std::barrier`. The above would create a separate `syclex::barrier` for every work-item, and store those barriers in private memory. 

I also was expecting the usage model above. Can you elaborate more on why the barrier is better declared outside the kernel ? For example, joint_matrix<sub_group/work_group, ..> will be declared inside the kernel, and that doesn't mean it is private per work-item. The scope defines the allocation.

Reply:
joint matrix is a "joint" type, so it is shared across the different WIs in the group.
I understand barrier is different and we want to keep it as close as std::barrier. but the initialization will make us diverge from std::barrier anyway.
The declaration placement makes sense but I was inquiring about the initialization that should happen separately. 
Are you preferring:
1- *mbar = syclex::barrier<sycl::memory_scope::device>(nthreads);
or 2-  placement new 
over 3- introducing a new init function?

Reply:
It will be also useful to show the usage (allocation and initialization) with "group_local_memory"

Reply:
Staying aligned with `std::barrier` will make it easier to eventually support `std::barrier` directly in kernels.  It would be great if we could train developers to think of `syclex::barrier` as an optimized (scope-aware) version of `std::barrier`, without also having to teach them brand new usage models and concepts.

My snippet above is incorrect, and it would need to use placement new:
```c++
void* barmem = (void*) sycl::device_malloc(sizeof(syclex::barrier<sycl::memory_scope::device>), q);
syclex::barrier<sycl::memory_scope::device>* mybar;
q.parallel_for(..., [=](sycl::nd_item it) {

  if (/* This work-item was elected by the user to construct and initialize the barrier*/) {
    *mybar = new (barmem) syclex::barrier<sycl::memory_scope::device>(nthreads);
  }
  
  // Synchronize the work-items via some existing mechanism, to ensure that mbar has been initialized
  sycl::group_barrier(it.get_root_group());

  // Use the barrier
  mybar->arrive_and_wait();

  if (/* This work-item was elected by the user to destruct the barrier */) {
   mybar->~barrier();
  }
}).wait();
sycl::free(barmem, q);
```

I think this is very similar to what you would write if you were trying to initialize a `std::barrier` after some threads have been created.  Most examples don't show that, though -- they show the barrier being created by a thread prior to the parallel work, and the barrier object being passed to the kernel.  As Greg noted, we can probably do that with USM too:

```c++
void* barmem = (void*) sycl::device_malloc(sizeof(syclex::barrier<sycl::memory_scope::device>), q);
auto* mybar = new (barmem) syclex::barrier<sycl::memory_scope::device>(nthreads);
q.parallel_for(..., [=](sycl::nd_item it) {
  
  // Use the barrier
  mybar->arrive_and_wait();

}).wait();
mybar->~barrier();
sycl::free(barmem, q);
```

The common use-case with `group_local_memory` is nice because it effectively handles all the barrier synchronization and placement new stuff internally, so I think it should be possible to write something like:

```c++
q.parallel_for(..., [=](sycl::nd_item it) {
  auto* gbar = syclex::group_local_memory<syclex::barrier<sycl::memory_scope::work_group>(it.get_group(), nthreads);
  gbar->arrive_and_wait();
}).wait();
```

If people really don't like the way that placement new works, then I'm not opposed to the idea of some sort of free function that would encapsulate this pattern.

I'm fairly certain that this is how developers migrating from `std::barrier` would expect things to work, but I've only written a few simple C++20 examples.  If we have reason to believe that `std::barrier` doesn't work this way, and that developers will expect to be able to define barrier objects directly inside of their kernels, please let me know and I'll revisit the design of this class.

Reply:
My opinion: If we can divert a bit from std::barrier and think of syclex::barrier as a joint data type (similar to joint_matrix<work_group,..>), then I think that would be much simpler. If I were a CUDA developer moving to SYCL, I would expect it to be defined inside the kernel. 

Also, in a typical DMA/MME kernel, we would have many barriers for double-buffering, producer-consumer threads sync, engine-to-engine sync (FS-AI). The code may get complicated.

Would we have issues with non-SYCL C++ compiler if we define barrier objects inside lambdas ? If yes, then how will we manage declarations of joint_matrix<work_group, ..> inside the kernel ? I assume each private object will be a wrapper around a pointer to common SLM location ? Can we do the same for syclex::barrier ?

Reply:
> If I were a CUDA developer moving to SYCL, I would expect it to be defined inside the kernel.

Can you say more about this?  My understanding is that `cuda::barrier` is similarly modeled on `std::barrier`, so I wouldn't expect their barriers to be defined inside of kernels either.  Note that barriers defined only for the work-group (in work-group local memory) are a special case -- for CUDA I'd expect them to be declared in global `__shared__` variables, and for SYCL I'd expect them to use `group_local_memory`.

> Also, in a typical DMA/MME kernel, we would have many barriers for double-buffering, producer-consumer threads sync, engine-to-engine sync (FS-AI). The code may get complicated.

I agree the complexity is something we need to fix.  I'd like to see if there are any ways to address the complexity before making barriers into "joint" objects.  I don't think there is anything _wrong_ with the mental model I sketched above, beyond that it's verbose and complex.  If we can identify the right way to implement things (the long way), we can add convenience functions to encapsulate those patterns.

> Would we have issues with non-SYCL C++ compiler if we define barrier objects inside lambdas ? If yes, then how will we manage declarations of joint_matrix<work_group, ..> inside the kernel ? I assume each private object will be a wrapper around a pointer to common SLM location ? Can we do the same for syclex::barrier ?

If you're asking about providing a fallback host implementation using only C++ features, I'm not sure.  I think it might be possible to implement things the way you describe, but I suspect it would be complicated to figure out how to map each of the barrier variables created by different work-items to the same memory location(s).  That was the issue we encountered when trying to implement `group_local_memory`, anyway.

That said, enabling an implementation of this feature for non-SYCL C++ compilers is not a goal of the extension.

Reply:
> Could you construct the barrier object on the host in USM shared memory, and then use the barrier in a kernel to synchronize work-items?

I think that would likely not work for hardware implementations. For example CUDA mbarrier uses a special PTX instruction to initialize the mbarrier and gives no instructions on how to initialize it on the host.

Reply:
> I think that would likely not work for hardware implementations. For example CUDA mbarrier uses a special PTX instruction to initialize the mbarrier and gives no instructions on how to initialize it on the host.

I think this is scope-dependent.  My understanding is that CUDA's `mbarrier` can only be used for barriers at work-group scope (and that the mbarrier's state is allocated in CUDA shared memory).  My expectation is that a barrier at device or system scope probably has to fall back to an implementation based on atomics, in which case it should be safe to initialize on the host or device.

That said, I can imagine somebody in future wanting to provide a hardware implementation of a device-scope barrier, too.  Maybe we should say that:
- work-group scope barriers have to be constructed by a work-item in the work-group that will use the barrier
- device-scope barriers have to be to be constructed on the device that will use the barrier
- system-scope barriers can be constructed on the host or device

Reply:
> work-group scope barriers have to be constructed by a work-item in the work-group that will use the barrier

We can say that to accommodate a potential more restrictive HW implementation. However, for CUDA I think it is enough to say it needs to be initialized on the device. I just noticed CUDA mbarrier also has another restriction compared to this proposal. It needs to be in local memory.

Reply:
I've added some draft wording restricting where the constructor can execute in 1870ab70b0cbbde9677aaa506bbad66544d67e73.

Although the PTX `mbarrier` object must be allocated in shared memory, `cuda::barrier` doesn't seem to have this restriction.  I'd rather not introduce this as a restriction until we definitely know that we need it.  Maybe there's something smart we can do on the implementation side, to use `mbarrier` if the `syclex::barrier` is allocated in group-local memory, and fall back to atomics otherwise?

If we can't figure out how to do that, then we can revisit this text.  We definitely need to be able to implement `syclex::barrier` using a PTX `mbarrier` when possible.